validation: Reduce direct g_chainman usage

dongcarl opened the pull request:
This PR paves the way for de-globalizing `g_chainman` entirely by removing the usage of `g_chainman` in the following functions/methods:
- `~CMainCleanup`
- `CChainState::FlushStateToDisk`
- `UnloadBlockIndex`

The remaining direct uses of `g_chainman` are as follows:
1. In initialization codepaths:
	- `AppTests`
	- `AppInitMain`
	- `TestingSetup::TestingSetup`
2. `::ChainstateActive`
3. `LookupBlockIndex`
	- Note: `LookupBlockIndex` is used extensively throughout the codebase and require a much larger set of changes, therefore I've left it out of this initial PR

dongcarl committed:
validation: Pass in chainman to UnloadBlockIndex

dongcarl committed:
validation: Move PruneOneBlockFile to BlockManager

[META] This is a pure refactor commit.

Move PruneBlockFile to BlockManager because:
1. PruneOneBlockFile only acts on BlockManager
2. Eliminates the need for callers (FindFilesToPrune{,Manual}) to have a
   reference to the larger ChainstateManager, just a reference to
   BlockManager is enough. See following commits.

dongcarl committed:
validation: Move FindFilesToPrune{,Manual} to BlockManager

[META] No behaviour change is intended in this commit.
[META] This commit should be followed up by removing the comments and
       assertions meant only to show that the change is correct.

Also stop FindFilesToPrune{,Manual} from unnecessary reaching for
::ChainActive() by passing in the necessary information.

dongcarl committed:
validation: Move ~CMainCleanup logic to ~BlockManager

~CMainCleanup:
1. Is vestigial
2. References the g_chainman global (we should minimize g_chainman refs)
3. Only acts on g_chainman.m_blockman
4. Does the same thing as BlockManager::Unload

MarcoFalke commented on src/init.cpp:
```diff
@@ -1559,7 +1559,7 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
                 chainman.m_total_coinsdb_cache = nCoinDBCache;
 
-                UnloadBlockIndex(node.mempool.get());
+                UnloadBlockIndex(node.mempool.get(), &chainman);
```
in commit bc872e44e29f7108720072a7d37ce06fb069ba4f: Doesn't it make more sense to pass it as `this` (i.e. make it a member function)?

I've done that in #19909 (commit facd88a77189f6953d28a6743e0c68175bc18f9e)

Conceptually it makes little sense for chainman to be optional. Why would you want to run a node with validation disabled? So, if you don't like to make it a member function, I'd say to at least pass it as reference.

dongcarl replied:
I will take another look at your PR, but I think in this particular instance, we should pass `chainman` as a reference, simply because `UnloadBlockIndex` touches so many validation globals that are not owned by `chainman`. Perhaps in the future when we move those globals into `chainman` or some other validation-container class, it would make more sense to do so.

I was worried about callers (to `UnloadBlockIndex`) who only have a pointer to `chainman`, but the only caller I see that only has a pointer to `chainman` is `~TestingSetup`, and it already looks like it makes the assumption that `m_node.chainman` is not null (see the `m_node.chainman->Reset()` line in `~TestingSetup`). Will make the change and take a look at your PR too.

jnewbery replied:
I also have a gut feeling that UnloadBlockIndex should eventually be a member function of chainman, but agree that we don't need to do everything in one PR.

maflcko reviewed (commented):
Concept ACK. I did similar changes or was planning to do them.

jamesob commented:
Concept ACK; helps reduce globals, supports fuzzing.

fanquake commented:
Concept ACK

jnewbery commented on src/validation.h:
```diff
@@ -407,6 +407,12 @@ class BlockManager {
+    //! Mark one block file as pruned (modify associated database entries)
+    void PruneOneBlockFile(const int fileNumber) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+
+    // See definition for documentation
```
Move the doxygen comments here (this comment made sense when it was next to a forward declaration, but now that the declaration is in the header file, it should have the doxygen comment)

dongcarl replied:
Resolved in cdd2063bbe8b8df4db42c81ee7913e2dc0b2387a

jnewbery commented on src/validation.h:
```diff
@@ -914,6 +917,15 @@ class ChainstateManager
     //! ResizeCoinsCaches() as needed.
     void MaybeRebalanceCaches() EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
+
+    ~ChainstateManager() {
```
Did you consider moving this logic to the `BlockManager` dtor instead of here? There's only ever one `BlockManager` instance (inside `ChainstateManager`), so destruction will happen at the same time. All the logic in this function is on `BlockManager` data.

jnewbery replied:
Actually, can't you just call `BlockManager::Unload()` in the `BlockManager` dtor?

dongcarl replied:
Resolved in 4668ded6d6ea4299d998abbb57543f37519812e2

jnewbery commented on src/validation.h:
```diff
@@ -914,6 +917,15 @@ class ChainstateManager
     void MaybeRebalanceCaches() EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
+
+    ~ChainstateManager() {
+        // block headers
```
Remove useless comment.

jnewbery commented on src/validation.h:
```diff
@@ -914,6 +917,15 @@ class ChainstateManager
+
+    ~ChainstateManager() {
+        // block headers
+        BlockMap::iterator it1 = BlockIndex().begin();
+        for (; it1 != BlockIndex().end(); it1++) {
```
```suggestion
        for (auto it = BlockIndex().begin; it != BlockIndex().end(); ++it) {
```

jnewbery reviewed (commented):
Concept ACK.

I think you may as well clean up the code as you're moving it. No need to retain comments that are no longer useful, for example.

Empact commented:
I get this warning: 

```
qt/test/apptests.cpp:87:45: warning: passing variable 'g_chainman' by reference requires holding mutex 'cs_main' [-Wthread-safety-reference]
    UnloadBlockIndex(/* mempool */ nullptr, g_chainman);
                                            ^
```

Proposed change: pass as a ref instead. Note that `chainman` is dereferenced 4 lines later in `setup_common.cpp`
```diff
diff --git a/src/init.cpp b/src/init.cpp
index 220ca061f2..90bf3c6461 100644
--- a/src/init.cpp
+++ b/src/init.cpp
diff --git a/src/init.cpp b/src/init.cpp
index 220ca061f2..90bf3c6461 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -1559,7 +1559,7 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
                 chainman.m_total_coinstip_cache = nCoinCacheUsage;
                 chainman.m_total_coinsdb_cache = nCoinDBCache;
 
-                UnloadBlockIndex(node.mempool.get(), &chainman);
+                UnloadBlockIndex(node.mempool.get(), chainman);
 
                 // new CBlockTreeDB tries to delete the existing file, which
                 // fails if it's still open from the previous loop. Close it first:
diff --git a/src/qt/test/apptests.cpp b/src/qt/test/apptests.cpp
index 32ec1ff475..0a4ae16589 100644
--- a/src/qt/test/apptests.cpp
+++ b/src/qt/test/apptests.cpp
@@ -84,7 +84,7 @@ void AppTests::appTests()
     // Reset global state to avoid interfering with later tests.
     LogInstance().DisconnectTestLogger();
     AbortShutdown();
-    UnloadBlockIndex(/* mempool */ nullptr, &g_chainman);
+    WITH_LOCK(::cs_main, UnloadBlockIndex(/* mempool */ nullptr, g_chainman));
     WITH_LOCK(::cs_main, g_chainman.Reset());
 }
 
diff --git a/src/test/util/setup_common.cpp b/src/test/util/setup_common.cpp
index 2e24ad1e20..0352d1d5e2 100644
--- a/src/test/util/setup_common.cpp
+++ b/src/test/util/setup_common.cpp
@@ -187,7 +187,7 @@ TestingSetup::~TestingSetup()
     m_node.connman.reset();
     m_node.banman.reset();
     m_node.args = nullptr;
-    UnloadBlockIndex(m_node.mempool.get(), m_node.chainman);
+    UnloadBlockIndex(m_node.mempool.get(), *(m_node.chainman));
     m_node.mempool.reset();
     m_node.scheduler.reset();
     m_node.chainman->Reset();
diff --git a/src/validation.cpp b/src/validation.cpp
index 774b6d435d..7da648ddea 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -4642,10 +4642,10 @@ void CChainState::UnloadBlockIndex() {
 // May NOT be used after any connections are up as much
 // of the peer-processing logic assumes a consistent
 // block index state
-void UnloadBlockIndex(CTxMemPool* mempool, ChainstateManager* chainman)
+void UnloadBlockIndex(CTxMemPool* mempool, ChainstateManager& chainman)
 {
     LOCK(cs_main);
-    if (chainman) chainman->Unload();
+    chainman.Unload();
     pindexBestInvalid = nullptr;
     pindexBestHeader = nullptr;
     if (mempool) mempool->clear();
diff --git a/src/validation.h b/src/validation.h
index 4c0555e387..1f90dbdaf2 100644
--- a/src/validation.h
+++ b/src/validation.h
@@ -159,7 +159,7 @@ void LoadExternalBlockFile(const CChainParams& chainparams, FILE* fileIn, FlatFi
 /** Ensures we have a genesis block in the block tree, possibly writing one to disk. */
 bool LoadGenesisBlock(const CChainParams& chainparams);
 /** Unload database information */
-void UnloadBlockIndex(CTxMemPool* mempool, ChainstateManager* chainman);
+void UnloadBlockIndex(CTxMemPool* mempool, ChainstateManager& chainman);
 /** Run an instance of the script checking thread */
 void ThreadScriptCheck(int worker_num);
 /**
```

dongcarl committed:
docs: Move FindFilesToPrune{,Manual} doxygen comment

[META] This is a pure comment commit.

They belong in the member declarations in the header file.

dongcarl commented:
Thank you MarcoFalke, jnewbery, and Empact for the code review!

Marco: `ChainstateManager` is now passed in as a reference in 577ec184bd47b3e88a1d3e60b4d2045bd51328b4
jnewbery: Thanks for the detailed review! See the review comments for resolutions
Empact: Good find. I'm turning on `-Werror=thread-safety` by default now and that particular problem is fixed in 577ec184bd47b3e88a1d3e60b4d2045bd51328b4

-----

Note: I will push a commit to remove the comments and assertions in 84d983656f1927955e516a94366b7c216bacded3 once I get the first Code Review ACK.

jnewbery commented on src/validation.cpp:
```diff
@@ -3957,22 +3993,22 @@ void UnlinkPrunedFiles(const std::set<int>& setFilesToPrune)
+    if (chain_tip_height < 0)
         return;
 
+
```
style: no need for extra line break.

dongcarl replied:
Resolved in fe07ba16511bd119b6ed744e68adfb3210a84254

jnewbery commented on src/validation.cpp:
```diff
@@ -3957,22 +3993,22 @@ void UnlinkPrunedFiles(const std::set<int>& setFilesToPrune)
 
     LOCK2(cs_main, cs_LastBlockFile);
-    if (::ChainActive().Tip() == nullptr)
+    if (chain_tip_height < 0)
```
style: when touching lines, fix code to match style guide (here, join the line below with this or add braces)

dongcarl replied:
Resolved in 597ec8a853df20fc68892b5f1131d27a1f2cf0e5

jnewbery reviewed (commented):
utACK cdd2063bbe8b8df4db42c81ee7913e2dc0b2387a

Just a couple of style suggestions inline. Feel free to ignore.

ryanofsky commented on src/validation.cpp:
```diff
@@ -2284,14 +2281,53 @@ bool CChainState::FlushStateToDisk(
+            // subsequent commit, as they're just meant to demonstrate
+            // correctness (you can run tests against it and see that nothing
+            // exit unexpectedly).
+            assert(std::addressof(::ChainActive()) == std::addressof(m_chain));
```
In commit "validation: Move FindFilesToPrune{,Manual} to BlockManager" (84d983656f1927955e516a94366b7c216bacded3)

Nice way of verifying, and didn't know about std::addressof!

ryanofsky reviewed (approved):
Code review ACK cdd2063bbe8b8df4db42c81ee7913e2dc0b2387a. Seems good and can re-ack if an extra assert-cleanup commit will be added.

dongcarl committed:
style: Make FindFilesToPrune{,Manual} match style guide

[META] This is a pure style commit.

dongcarl committed:
validation: Remove review-only comments + assertions

[META] This is a followup to "validation: Move FindFilesToPrune{,Manual}
       to BlockManager" removing comments and assertions meant only to
       show that the change is correct.

dongcarl commented on src/qt/test/apptests.cpp:
```diff
@@ -84,8 +84,11 @@ void AppTests::appTests()
-    WITH_LOCK(::cs_main, g_chainman.Reset());
+    {
+        LOCK(cs_main);
+        UnloadBlockIndex(/* mempool */ nullptr, g_chainman);
```
Resolved in 74f73c783d46b012f375d819e2cd09c792820cd5 (I just removed that comment since "mostly pure refactor" isn't too meaningful :smile: 

For the `[META]` tags, I literally just came up with it randomly, mostly to describe high-level things that apply to the entire commit or things that reference other commits.

dongcarl commented:
Got a [code review ACK](https://github.com/bitcoin/bitcoin/pull/19927#pullrequestreview-488777314), just pushed a version which:
1. Has a [last commit](75142a7bd7890f2e14920dbe9701ada2cbc2326f) which removes the review-only comments + assertions
2. Adds a [style-only commit](597ec8a853df20fc68892b5f1131d27a1f2cf0e5) which

jnewbery commented:
code review ACK 75142a7bd7890f2e14920dbe9701ada2cbc2326f

ryanofsky reviewed (approved):
Code review ACK 75142a7bd7890f2e14920dbe9701ada2cbc2326f. Since last review just new commit removing asserts, new commit cleaning up whitespace, and a newline fix in earlier commit

MarcoFalke commented:
Previously those methods were private (`static` in a cpp file), so I am a bit worried about exposing them as public member functions to all of Bitcoin Core (rpc, net_processing, ...). 

What do you think about turning them `private` again?

E.g.

```diff
diff --git a/src/validation.h b/src/validation.h
index cc5e36acfc..bc9b136039 100644
--- a/src/validation.h
+++ b/src/validation.h
@@ -357,7 +357,12 @@ struct CBlockIndexWorkComparator
  * This data is used mostly in `CChainState` - information about, e.g.,
  * candidate tips is not maintained here.
  */
-class BlockManager {
+class BlockManager
+{
+    friend CChainState;
+    void FindFilesToPruneManual(std::set<int>& setFilesToPrune, int nManualPruneHeight, int chain_tip_height);
+    void FindFilesToPrune(std::set<int>& setFilesToPrune, uint64_t nPruneAfterHeight, int chain_tip_height, bool is_ibd);
+
 public:
     BlockMap m_block_index GUARDED_BY(cs_main);
 
@@ -411,24 +416,6 @@ public:
     //! Mark one block file as pruned (modify associated database entries)
     void PruneOneBlockFile(const int fileNumber) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
-    /* Calculate the block/rev files to delete based on height specified by user with RPC command pruneblockchain */
-    void FindFilesToPruneManual(std::set<int>& setFilesToPrune, int nManualPruneHeight, int chain_tip_height);
-    /**
-     * Prune block and undo files (blk???.dat and undo???.dat) so that the disk space used is less than a user-defined target.
-     * The user sets the target (in MB) on the command line or in config file.  This will be run on startup and whenever new
-     * space is allocated in a block or undo file, staying below the target. Changing back to unpruned requires a reindex
-     * (which in this case means the blockchain must be re-downloaded.)
-     *
-     * Pruning functions are called from FlushStateToDisk when the global fCheckForPruning flag has been set.
-     * Block and undo files are deleted in lock-step (when blk00003.dat is deleted, so is rev00003.dat.)
-     * Pruning cannot take place until the longest chain is at least a certain length (100000 on mainnet, 1000 on testnet, 1000 on regtest).
-     * Pruning will never delete a block within a defined distance (currently 288) from the active chain's tip.
-     * The block index is updated by unsetting HAVE_DATA and HAVE_UNDO for any blocks that were stored in the deleted files.
-     * A db flag records the fact that at least some block files have been pruned.
-     *
-     * @param[out]   setFilesToPrune   The set of file indices that can be unlinked will be returned
-     */
-    void FindFilesToPrune(std::set<int>& setFilesToPrune, uint64_t nPruneAfterHeight, int chain_tip_height, bool is_ibd);
     /**
      * If a block header hasn't already been seen, call CheckBlockHeader on it, ensure
      * that it doesn't descend from an invalid block, and then add it to m_block_index.

MarcoFalke commented:
cr ACK 75142a7bd7890f2e14920dbe9701ada2cbc2326f 😺

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

cr ACK 75142a7bd7890f2e14920dbe9701ada2cbc2326f 😺
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhyqwv/eNXZQ6DzY+d3CX83lRRCFbjD3zfjQ0o0dT2SBGnP4oaG6gI+fwzUQacl
ZdtAZ3JEhJU6RDFs+gMun19ldVP7SG+RhJAGHRZgNuWzT0VvC/I1USyK/QZ7Rb53
G2Rz7oSIMTZ9X88kVBuBVYPphSalSVKDTwbPoV5tYUn96Z753aG768FS1pop6Lvx
LwFLLVWjN8hpA1UAinh57qIAy9Jr21jA1bGcUG5M8AydqTlwA1pTig80Yktn5cvb
Me9FWSLHqifUHR2Js8TUjwykChul18d/8C9C4kmOynDeH4b+n7nmYFP1Ynrh458X
P71pSb0vcVFK1fXnFFVh3z493Is2vpGxxuZino4RPvK4+53G7fJ41xlL78r+2cRk
gKrBAukJNMM02RRwdQ1W+4jSpcZcf8PHLUNSg7BVOP5AFLcCekvhExVLIJJO9LiS
Q5jNu4OQSVEuoZhjbmoeHQhgPe+lrd3LYWvPtpJveos24+gTNSS3KWFD1IB3QvzK
HIpk9O5/
=izRp
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `3230012495f1048f1411855fc06225652a1358399085dc6b400e304c7f82dadb  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401083230012495f1048f1411855fc06225652a1358399085dc6b400e304c7f82dadbf01013255e7a96728bed9b4f3c17ee1ce96008fff010057b6b712b35e5da800f897e8b9e753a08f1045f670e6ff008d9dec54e95a7ad710083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0104d400ed7649431e136461dee16a397c808f1045f670e6df0086215af5b970539220083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0104fe0a6d7756d26a1ad8de59d3e0f835f08f1045f670e6df008fb8f52995c50205e0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df0107c2543e58756487877c7da29d6ab3c5908f1045f670e6ff0082c828ef742ef5b560083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

dongcarl commented:
Made `FindFilesToPrune{,Manual}` private per Marco's [comment](https://github.com/bitcoin/bitcoin/pull/19927#issuecomment-695758685). This can be reviewed with:

```sh
git diff 75142a7 72a1d5c
```

jnewbery commented on src/validation.h:
```diff
@@ -356,7 +356,31 @@ struct CBlockIndexWorkComparator
+    friend CChainState;
+
+private:
+    /* Calculate the block/rev files to delete based on height specified by user with RPC command pruneblockchain */
```
(Only if you retouch the branch again) Make this a doxygen comment (start with `/**`). Also line-wrap this comment and the one below at something sensible :slightly_smiling_face:

jnewbery reviewed (commented):
utACK 72a1d5c6f3834e206719ee5121df7727aed5b786

One style suggestion inline. Only consider it if you retouch the branch.

MarcoFalke commented:
re-ACK 72a1d5c6f3 👚

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 72a1d5c6f3 👚
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhTrAv+Ky3GNUn0kO04zX+7dsU23mPYIvAUJ/SRXBtSQID9lNph+93gmf29iMWM
3eOnDT1KhSINP5mHGRX9j4loYTh2VxUE9Hwqzc/em4bfh+QcE/Fyjxx6uArGVZBQ
cYCEwsoQ51nbp9/Yftt2UFQzjjtHXtEGn7KO++u2WDzpOtL8vdxOZTK+e0Xb6Gbd
XJSDGf2D6YeAhElyMETc6TJQ3/E7nHTWurNjXVR/C6EbeeVzrTlaLmGVGUBqWcAf
qcQjVHB3qrQHrXfr+bK6XlSOJ4/60T+Q801mX2AeUK4ygpgcsVpWquaviC1DqzfM
kdrdssVPW9b5yuUM8ciaPxLTiXBqyzKVaUA8Q3CADsQClStA7MOfUItmo9HUs/Lz
VWughvafPFfI+cbHkR3rvvvMcPza12/9sYYfCi+xHfw1CaMnjs5lvDD0Cw4ShGMH
h/vJzV/MFnDquVrg/T1Cam1zkXf3bxGsc7K/mJh3wO99Kgf9X9nlS7Y77eQIeWx2
bzQoIo/G
=9glt
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `bcdea3595c56b92fa07e06e3a4cdb3a9a7439cb78f42e7fcdf109c35f3d1dd79  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108bcdea3595c56b92fa07e06e3a4cdb3a9a7439cb78f42e7fcdf109c35f3d1dd79f0108ff9d29c43c3d6dc694e93253a4531cc08fff0106dae1a5708a642bbde08f35182544b7908f1045f6b952bf0085abce0ed22949f0e0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010c306f28c43b05661189d352bbe487e0008f1045f6b952cf008329198c87a3d41430083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>
